### PR TITLE
Implement maximum_likelihood function

### PR DESCRIPTION
**Pull request by Ayush-iitkgp**
This the first implementation of the likelihood function. The objective is to share the code and ask for reviews and improve upon it.
Issue #29

### CODE REVIEW

File: src/build_maximum_likelihood_objective.jl [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  maximum_likelihood_function = function (p)
+    if verbose_opt
+      count::Int += 1
+      if mod(count,verbose_steps) == 0
+        println("f_$count($p)")
+      end
+    end
+    tmp_prob = prob_generator(prob,p)
+    sol = solve(tmp_prob,alg;saveat=likelihood.t,save_everystep=false,dense=false,kwargs...)
+    likelihood(sol)
```
Comment (ChrisRackauckas) [inline src/build_maximum_likelihood_objective.jl]: Why is this a different outer function? The likelihood is just a different cost function, so this should be unnecessary.

File: src/cost_functions.jl [hunk 2]
```diff
@@ -42,3 +42,25 @@ end
...
+function (f::MaximumLikelihood)(sol::DESolution)
+  data = f.data
+  variance = f.variance
+  fill_length = length(f.t)-length(sol)
+  for i in 1:fill_length
+    push!(sol.u,fill(Inf,size(sol[1])))
+  end
+  prod = 1.0
+  @inbounds for i in 1:length(sol)
+    for j in 1:length(sol[i])
```
Comment (ChrisRackauckas) [inline src/cost_functions.jl]: you can make double-loops in one line

File: src/cost_functions.jl [hunk 3]
```diff
@@ -42,3 +42,25 @@ end
...
+end
+
+function (f::MaximumLikelihood)(sol::DESolution)
+  data = f.data
+  variance = f.variance
+  fill_length = length(f.t)-length(sol)
+  for i in 1:fill_length
+    push!(sol.u,fill(Inf,size(sol[1])))
+  end
+  prod = 1.0
```
Comment (ChrisRackauckas) [inline src/cost_functions.jl]: make this match the type of the solution, i.e. `one(eltype(sol.u))`

File: src/cost_functions.jl (hunk 2 above)
Comment (ChrisRackauckas) [inline src/cost_functions.jl]: for i in 1:length(sol), j in 1:length(sol[i])

File: src/cost_functions.jl [hunk 4]
```diff
@@ -42,3 +42,25 @@ end
...
+  @inbounds for i in 1:length(sol)
+    for j in 1:length(sol[i])
+      prod *= exp(((data[j,i] - sol[j,i])^2)/(-2*variance[j,i]))/(sqrt(2*π*variance[j,i]))
```
Comment (ChrisRackauckas) [inline src/cost_functions.jl]: Does the compiler elide the 2*π or compute it each round? You might want to save that as a constant. Also, this line is a very good candidate for an `@fastmath` annotation to speed up the `exp` and `sqrt`. The extra error essentially acts as regularization anyways.

**Comment by ChrisRackauckas** [comment]
Made some style points for the implementation, but other than those I think it's pretty good. Do you have tests on the basic Lotka-Volterra problem yet?

File: src/build_maximum_likelihood_objective.jl (hunk 1 above)
Comment (Ayush-iitkgp) [inline src/build_maximum_likelihood_objective.jl]: You mean to replace `likelihood` by `loss`?

Comment (ChrisRackauckas) [inline src/build_maximum_likelihood_objective.jl]: The only difference between the two functions is you changed the name of the loss function. Won't the likelihood function work with the loss objective? There's no reason to duplicate code to change a few names.

**Comment by ChrisRackauckas** [comment]
From what I can see, `variance` is not used as a parameter, and so it will not be optimized. Don't you want to treat those as non-constant hyperparameters to train along with the other parameters?

File: src/build_maximum_likelihood_objective.jl (hunk 1 above)
Comment (Ayush-iitkgp) [inline src/build_maximum_likelihood_objective.jl]: `loss` takes 3 or 2 arguments(data,t and optional loss_function) whereas likelihood takes 3 arguments (data, t, and variance of the population from which the data has been drawn).
Does that make sense to give it a different name?

Comment (ChrisRackauckas) [inline src/build_maximum_likelihood_objective.jl]: Their constructors are different, but their calls are the same.

**Comment by Ayush-iitkgp** [comment]
No, the user has to pass the `variance` as the argument to build_maximum_likelihood_objective https://github.com/csynbiosys/AMIGO2R2016b/blob/master/Examples/ParameterEstimation_PE/NFKB_PE.m#L141

**Comment by ChrisRackauckas** [comment]
>No, the user has to pass the variance as the argument to build_maximum_likelihood_objective 

Then why don't you do that? You're having it passed to the likelihood construction instead. I agree that it needs to be passed in in order to be optimized as noted earlier. But you can just recognize that it's a likelihood function via some abstract type and then use that to append the variances as part of the parameters.

**Comment by Ayush-iitkgp** [comment]
This is the reason why I named it `likelihood` instead of `loss`. Just to differentiate the definitions.
You want it to be explicitly passed to build_maximum_likelihhod_objective?

**Comment by ChrisRackauckas** [comment]
Either way is fine. You can either check if it's a likelihood function and grab the initial variances, or add that they can be passed in. Either way, there's one spot that handles the variances, but the rest should be exactly the same, right?

Also, don't look at the AMIGO2 code. It's GPL licensed. You're breaking the law if any of your code is derivative of it. You can use their paper as a reference, but do not use their code as a reference.

**Comment by ChrisRackauckas** [comment]
So why exactly does `build_maximum_likelihood_objective` have to be a different function than `build_loss_objective`? It does everything the same?

**Comment by Ayush-iitkgp** [comment]
The most recent commit also addresses regularization issue #33

**Comment by Ayush-iitkgp** [comment]
@ChrisRackauckas if you are satisfied, I will modify the tests accordingly.

**Comment by Ayush-iitkgp** [comment]
The new API would be:
```
W = eye(sizeof(p))
penalty = MahalanobisPenalty(W)
lambda = 0
cost_function = build_loss_objective(prob,Tsit5(),CostVData(t,data),Regularization(lambda,penalty),maxiters=10000)
```
lambda = 0 implies no regularization

**Comment by ChrisRackauckas** [comment]
`lambda = 0 implies no regularization`

Why not just have it as a default argument:

```julia
function build_loss_objective(prob::DEProblem,alg,loss,regularization = nothing
```

This way, if the user doesn't pass it, it becomes type information and you can compile away that branch. Then:

```julia
if regularization == nothing
  return loss(sol)
else 
  return loss(sol) + regularization(p)
end
```

will compute which of the two it should do when the arguments are passed in.

**Comment by ChrisRackauckas** [comment]
Yeah, other than that style point, looks good.

File: src/build_loss_objective.jl [hunk 5]
```diff
@@ -8,7 +8,7 @@ end
 (f::DiffEqObjective)(x) = f.cost_function(x)
 (f::DiffEqObjective)(x,y) = f.cost_function2(x,y)
 
-function build_loss_objective(prob::DEProblem,alg,loss,regularization;mpg_autodiff = false,
+function build_loss_objective(prob::DEProblem,alg,loss;regularization=nothing,mpg_autodiff = false,
```
Comment (ChrisRackauckas) [inline src/build_loss_objective.jl]: default argument, not a keyword argument. `build_loss_objective(prob::DEProblem,alg,loss,regularization=nothing;`, see the difference with the `;`. Keyword arguments won't dispatch so that'll be somewhat of an issue (though actually not that big of a deal). But this will feel more natural, i.e. you usually leave it off but add it when you want.

**Comment by Ayush-iitkgp** [comment]
@ChrisRackauckas the maximum_likelihood_cost_function is taking forever to converge with NLopt and very poor convergence for Genetic Algorithm (0.72). Would you also want to look into it?

**Comment by ChrisRackauckas** [comment]
Forever in terms of time or number of iterations?

**Comment by Ayush-iitkgp** [comment]
In terms of time.

**Comment by ChrisRackauckas** [comment]
But the number of iterations are similar? So it's a code optimization problem?

**Comment by Ayush-iitkgp** [comment]
I think it's the problem with the complexity of the likelihood function given that it is exponential. In fact, this is the reason why we construct log-likelihood at the first stage.

**Comment by ChrisRackauckas** [comment]
But does it take the same number of iterations, just each iteration taking longer to compute?

**Comment by Ayush-iitkgp** [comment]
Nothing much has changed by negation. The maximum_likelihood_cost_function is still not converging. Are we doing anything wrong? @ChrisRackauckas @finmod any views?

**Comment by ChrisRackauckas** [comment]
Have you changed the `variance` around?
